pyiron_base/storage/helper_functions.py

```diff
@@ -40,3 +53,81 @@ def write_hdf5(
         at_most=10,
         delay=1,
     )
+
+
+def write_hdf5_with_json_support(value, path, file_handle):
+    value, use_json = _check_json_conversion(value=value)
+    try:
+        write_hdf5(
+            file_handle,
+            value,
+            title=path,
+            overwrite="update",
+            use_json=use_json,
+        )
+    except TypeError:
+        raise TypeError(
+            "Error saving {} (key {}): DataContainer doesn't support saving elements "
+            'of type "{}" to HDF!'.format(value, path, type(value))
+        ) from None
+
+
+def write_dict_to_hdf(hdf, data_dict, groups=[]):
+    with open_hdf5(hdf.file_name, mode="a") as store:
+        for k, v in data_dict.items():
+            if k not in groups:
+                write_hdf5_with_json_support(
+                    file_handle=store, value=v, path=hdf.get_h5_path(k)
+                )
+        for group in groups:
+            hdf_group = hdf.create_group(group)
+            for k, v in data_dict[group].items():
+                write_hdf5_with_json_support(
+                    file_handle=store, value=v, path=hdf_group.get_h5_path(k)
+                )
```

Comment:
I don't understand the purpose of `groups` or how it works.  `groups` only seems to refer top level sub dicts of `data_dict`, what about deeper nested ones?  At least clarify expected structure of `data_dict` in docstring.

Reply:
The issue is that the dictionary we get from the job is not stored in the HDF5 file as a single entry but rather can contain certain sub groups which we want to represent as individual groups in the HDF5 file. So far I introduced a single level of sub groups but you are right we can have an arbitrary level of hierarchy and the developers should be able to decide which entries they want to store individually so they can be reloaded more quickly. Or in terms of the HDF5 file which entries should be compressed as one JSON dictionary and which other entries are resolved as individual nodes on the HDF5 level.

Reply:
I thought about this a bit more and I think the best way to address this is to build a flat dictionary, where the keys represent the h5_path, something like this:
```
data_dict = {
    'job_name/input/control_inp/data_dict': {}, 
    'job_name/output/generic/energy_tot': [...],
    'job_name/output/generic/energy_pot': [...],
}
```